File: realm-common/src/main/java/io/github/solusmods/eternalcore/realm/impl/network/c2s/SyncRealmStoragePayload.java
package io.github.solusmods.eternalcore.realm.impl.network.c2s;

import dev.architectury.networking.NetworkManager;
import dev.architectury.utils.Env;
import io.github.solusmods.eternalcore.realm.EternalCoreRealm;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.server.level.ServerPlayer;

public record SyncRealmStoragePayload(CompoundTag data) implements CustomPacketPayload {
    public static final Type<SyncRealmStoragePayload> TYPE = new Type<>(EternalCoreRealm.create("sync_realms"));
    public static final StreamCodec<FriendlyByteBuf, SyncRealmStoragePayload> STREAM_CODEC = CustomPacketPayload.codec(SyncRealmStoragePayload::encode, SyncRealmStoragePayload::new);

    public SyncRealmStoragePayload(FriendlyByteBuf buf) {
        this(buf.readNbt());
    }

    public void encode(FriendlyByteBuf buf) {
        buf.writeNbt(data);
    }

    public void handle(NetworkManager.PacketContext context) {
        if (context.getEnvironment() != Env.SERVER) return;
        context.queue(() -> ClientAccess.handle(this, (ServerPlayer) context.getPlayer()));
    }

    /**
     * @return 
     */
    @Override
    public Type<? extends CustomPacketPayload> type() {
        return TYPE;
    }
}

File: realm-common/src/main/kotlin/io/github/solusmods/eternalcore/realm/impl/network/c2s/SyncRealmStoragePayload.kt
package io.github.solusmods.eternalcore.realm.impl.network.c2s

import dev.architectury.networking.NetworkManager
import dev.architectury.utils.Env
import io.github.solusmods.eternalcore.realm.EternalCoreRealm
import net.minecraft.nbt.CompoundTag
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.codec.StreamDecoder
import net.minecraft.network.codec.StreamMemberEncoder
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.server.level.ServerPlayer


data class SyncRealmStoragePayload(val data: CompoundTag?) : CustomPacketPayload {
    constructor(buf: FriendlyByteBuf) : this(buf.readNbt())

    fun encode(buf: FriendlyByteBuf) {
        buf.writeNbt(data)
    }

    fun handle(context: NetworkManager.PacketContext) {
        if (context.environment != Env.SERVER) return
        context.queue { ClientAccess.handle(this, context.player as ServerPlayer?) }
    }

    /**
     * @return
     */
    override fun type(): CustomPacketPayload.Type<out CustomPacketPayload?> {
        return TYPE
    }

    companion object {
        val TYPE: CustomPacketPayload.Type<SyncRealmStoragePayload?> =
            CustomPacketPayload.Type<SyncRealmStoragePayload?>(EternalCoreRealm.create("sync_realms"))
        val STREAM_CODEC: StreamCodec<FriendlyByteBuf?, SyncRealmStoragePayload?> =
            CustomPacketPayload.codec<FriendlyByteBuf?, SyncRealmStoragePayload?>(
                { obj: SyncRealmStoragePayload?, buf: FriendlyByteBuf? -> obj!!.encode(buf!!) },
                { buf: FriendlyByteBuf? ->
                    SyncRealmStoragePayload(
                        buf!!
                    )
                })
    }
}
